pycbc/workflow/core.py

```diff
@@ -198,7 +198,10 @@ def __init__(self, cp, name,
             raise TypeError("Failed to find %s executable " 
                             "at %s" % (name, exe_path))
         
-        self.add_pfn(exe_path)
+        if exe_path.startswith('gsiftp://'):
+            self.add_pfn(exe_path,site='pycbc-code')
```

Comment:
I agree with you Ian that the handling of the site property is a bit weird,
but here is my best understanding. Pegasus has no knowledge of where data
actually is and doesn't really associate them with "living" on a particular
site. It, by default, assumes that any url it has is valid from wherever it
tries to use it.

The site attribute really only does one thing.

1) If you request to run on a particular site, and your job needs data, it
will prefer PFNs that have your site attribute.

However this interacts badly with the force symlinks options.

2) Force symlinks turns *any* PFN into a symlink if you run on the same
site as the pool, and ignores the actual url structure, only looking at the
folder path.

To me, it is the symlinks option that is broken. It should only transform
file:// urls and fail elsewhere.


On Fri, Nov 6, 2015 at 10:57 AM, spxiwh <notifications@github.com> wrote:

> In pycbc/workflow/core.py
> <https://github.com/ligo-cbc/pycbc/pull/559#discussion_r44120432>:
>
> > @@ -198,7 +198,10 @@ def __init__(self, cp, name,
> >              raise TypeError("Failed to find %s executable "
> >                              "at %s" % (name, exe_path))
> >
> > -        self.add_pfn(exe_path)
> > +        if exe_path.startswith('gsiftp://'):
> > +            self.add_pfn(exe_path,site='pycbc-code')
>
> Just so as I can understand this better, because I never really understood
> why this needs a site= entry at all and we do use this elsewhere (ie.
> remote frame files if querying a non-local backup frame server). It's a
> gsiftp link, you get it by running globus-url-copy and that's it. Why do
> you need to know anything about the remote site that using a generic
> site="nonlocal" doesn't fulfill? Hell, using site="local" works in that the
> file will get gsiscp-ed, but then if using the symlink options pegasus will
> try to symlink against a location that doesn't exist, and pegasus *knows*
> doesn't exist (to me that seems a bug, and I don't understand why you/Karan
> say it is not). I think I don't understand the "site=" property for gsiftp
> links at all, it should be irrelevant, no?
>
> Using site='pycbc-code' implies only code.phy.syr.edu which seems
> restrictive.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/559/files#r44120432>.
>